### ISSUE TITLE
[2.0.0] Put the get prefixes back or not?

### ISSUE THREAD

**Issue by astei** [PaperMC/Velocity#501]
I consider this discussion to be a complete distraction, but I am opening this issue anyway to see if we can come to consensus on this.

_Please be civil._

## Background

In 47c354e6 I removed most of the `get` prefixes in the Velocity API. Then today, I created the [velocity-language-kotlin](https://github.com/VelocityPowered/velocity-language-kotlin) project. This sparked a vigorous not-a-debate on the Velocity Discord, which leads to this issue.

## Arguments for putting the prefix back

* **Kotlin has special support for it** - instead of writing, say, `proxyServer.eventManager()`, you can just write `proxyServer.eventManager` instead, saving developers a few keystrokes.
* **It reduces the delta for upgrading a plugin from Velocity 1.1.x** - since Velocity 1.1.x already largely used `get` prefixes, we could make the behavior more consistent in Velocity 2.0.0.
* **It conforms with the rest of Java** - this is a long-time established pattern in Java.
* **There is no need to change for change's sake** - if something works, better to stick with it. Getter prefixes have gotten us far.

## Arguments for the new status quo

* **The pattern is dated** - the Java platform has only recently (as of 2017) really began to start moving in a more modern direction. New Java features like the HTTP client and records completely omit `get` prefixes.
* **The change overall benefits more developers** - there are a vanishingly small number of Velocity plugins written in Kotlin, a language not common in the Minecraft community to begin with. A quick search on GitHub reveals fewer than 20 Velocity plugins written in Kotlin, compared to far more (I estimate over 150 or so) written in Java.
* **It's not consistent with other platforms** - Adventure 4, Sponge 8, Configurate 4, and all the methods Paper has added to support Adventure - all lack the `get` prefix.
* **Now is the best time to make the change** - we're not going to have an opportunity to make this change again for at least another year or so.

**Comment by ghost** [comment]
Still, it is better to use the absence of get/set, as it is correctly noted-the vast majority of Java developers.

Although Kotlin focuses on the get/set prefixes, it still provides an alternative in the form of extension-functions or value classes.

As an option for the entire Java / Kotlin community - raise the topic on the JetBrains issue tracker about the need to introduce new annotations in jetbrains-annotations such as @Getter @Setter just as it was with @Nullable and @NonNull. In addition, it is precisely from the annotations that null kotlin defines syntactic sugar in the form of the "?" operator.

**Comment by astei** [comment]
Useful prior discussion: SpongePowered/SpongeAPI#2276

**Comment by kennytv** [comment]
The "paradigm" and keystroke arguments are rather petty and in terms of moving on from dated format, it's always a bit of a "either now or never", which you only really have the opportunity for in major updates like this - otherwise it would only become more dated from now onwards

**Comment by Proximyst** [comment]
Although I am on the get/set side, as Kotlin is an argument for get/set, I will bring up Clojure as an argument against:

```clojure
(.getUniqueId player-here) ; doesn't look like an accessor
(.uniqueId player-here) ; looks and feels like an accessor
```

I believe this completely nullifies the argument, as there will always be another language that does/benefits from the opposite.

**Comment by A248** [comment]
My personal opinion is that get prefixes do not make sense for immutable objects, which have only getters. This is consistent with JDK 16 records.

However, in the presence of stateful mutation via setters, `propertyName(property)` is ambiguous, so in these cases, `get` and `set` provide useful clarify as to what the methods do.

As for `with` prefixes on immutable objects which return new objects, I'd consider that another issue altogether and possibly outside the scope of this one.

**Comment by Andre601** [comment]
I have to be this guy to ask this question but.... What about removing the `get` prefix is modern/more up to date?
I see seriously no real improvement in removing the get.

I myself even see this going against established standards, as it was pointed out in OP. Like when reading `MyClass.user()` is my first thought that this would just be accessing the `user` variable in the `MyClass` class (Even with the () visible can this confusion happen).
Meanwhile `MyClass.getUser()` makes it more clear. We GET the user info here through the dedicated method and not directly through the variable.

So tl;dr for me here could be, that the get prefix clarifies stuff more and prevents confusion about whether we just get a var directly or through a method.

While I do support (or to be honest "tolerate") Kotlin, do I feel like that the Java-base shouldn't need to (more or less) get forced to their standards...

**Comment by vectrixdevelops** [comment]
I would say in many situations the prefixes don't really help with much. Dropping the prefix doesn't reduce the readability in my opinion. If anything it helps a small amount by removing something that is essentially implied in most cases.

Obviously making this change does cause more friction for people wanting to update to the newer version, however, it could be argued to be minor in comparison to other breaking changes coming.

I personally would say stick with the direction of going without the prefixes.

**Comment by Andre601** [comment]
Mentioned this in the `#💻-velocity-development` channel but mention it here too.

When the decision with removing the get prefix is kept, should some methods get updated/renamed.
A prime example is `ProxyServer#getVersion()` (now `ProxyServer#version()`) which, when you try to get the String representing the proxy version, will result in the following code:
```java
ProxyServer#version().version();
```
This becomes kind of unreadable as it becomes more unclear what each `version()` does/returns. I don't say it was better with the get prefix, but I want to say that this should be taken as an opportunity to improve common naming conventions here.

My suggestions for this particular case here are as follows:

1. Rename first `version()` to `proxyVersion()` to match the class' name which this method returns.
2. Rename the `ProxyVersion` class to something like `ProxyInfo` and rename the method to something like `info()`

In general, should the version method get a renaming of some kind to reduce possible confusion to new developers and/or contributors.

**Comment by astei** [comment]
It's pretty clear that the prefixes are being put back.

**Comment by astei** [comment]
It's pretty clear that the prefixes are not being put back.
